content/doc/book/pipeline/jenkinsfile.adoc

```diff
@@ -889,6 +853,182 @@ within a single `withCredentials( ... ) { ... }` step by doing the following:
+
+```
+pipeline {
+    agent any
+    environment {
+        EXAMPLE_CREDS = credentials('example-credentials-id')
+    }
+    stages {
+        stage('Example') {
+            steps {
+                /* CORRECT */
+                sh('curl -u $EXAMPLE_CREDS_USR:$EXAMPLE_CREDS_PSW https://example.com/')
+            }
+        }
+    }
+}
+```
+
+
+==== Injection via interpolation
+
+[WARNING]
+======
+Groovy string interpolation can inject rogue commands into command interpreters via special characters.
+======
+
+Another note of caution. Using Groovy string interpolation for user-controlled variables with steps that pass their arguments to command interpreters such as the `sh`, `bat`, `powershell`, or `pwsh` steps can result in problems analogous to SQL injection.
+This occurs when a user-controller variable (generally an environment variable, usually a parameter passed to the build) that contains special characters (e.g. `' " / \ $ & % ^ > < | ;`) is passed to the `sh`, `bat`, `powershell`, or `pwsh` steps using Groovy interpolation.
+For a simple example:
+
+```
+pipeline {
+  agent any
+  parameters {
+    string(name: 'STATEMENT', defaultValue: 'hello; ls /', description: 'What should I say?')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        /* WRONG! */
+        sh("echo ${STATEMENT}")
+      }
+    }
+  }
+}
+```
+
+In this example, the argument to the `sh` step is evaluated by Groovy, and `STATEMENT` is interpolated directly into the argument as if `sh('echo hello; ls /')` has been written in the Pipeline.
+When this is processed on the agent, rather than echoing the value `hello; ls /`, it will echo `hello` then proceed to list the entire root directory of the agent.
+Any user able to control a variable interpolated by such a step would be able to make the `sh` step run arbitrary code on the agent.
+To avoid this problem, make sure arguments to steps such as `sh` or `bat` that reference parameters or other user-controller environment variables use single quotes to avoid Groovy interpolation.
+
+```
+pipeline {
+  agent any
+  parameters {
+    string(name: 'STATEMENT', defaultValue: 'hello; ls /', description: 'What should I say?')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        /* CORRECT */
+        sh('echo ${STATEMENT}')
+      }
+    }
+  }
+}
+```
+
+Credential mangling is another issue that can occur when credentials that contain special characters are passed to a step using Groovy interpolation.
+When the credential value is mangled, it is no longer valid and will no longer be masked in the console log.
+
+```
+pipeline {
+  agent any
+  stages {
+    stage('Example') {
+      steps {
+        withCredentials([string(credentialsId: 'secret-with-symbols', variable: 'SECRET_VALUE')]) {
+          // value of credential = sec%ret
+          /* WRONG! */
+          bat "echo %SECRET_VALUE%"
+        }
+      }
+    }
+  }
```

Comment:
In this case (just the wrong example, not the correct example), you have to use `$` instead of `%` for Groovy interpolation to take effect. Also, using `withCredentials` like this in Declarative is supported, but it would be clearer to use the Declarative-specific credentials syntax in `environment` like in your first two examples.

```suggestion
  environment {
        EXAMPLE_KEY = credentials('example-key') // Secret value is 'sec%ret'
  }
  stages {
    stage('Example') {
      steps {
        /* WRONG! */
        bat "echo $EXAMPLE_KEY"
      }
    }
  }
```